src/raft.rs

```diff
@@ -888,12 +1002,106 @@ impl<T: Storage> Raft<T> {
+        }
+        if use_leader_replication {
+            let core = &mut self.r;
+            let msgs = &mut self.msgs;
+            self.prs
+                .iter_mut()
+                .filter(|&(id, _)| *id != self_id)
+                .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+            return;
+        }
+
+        // If follower replication is enabled, MsgAppends sent to the same broadcast group
+        // will be merge into a MsgGroupBroadcast.
+        //
+        // Messages that needs to be forwarded are stored in hashmap temporarily,
+        // and they are grouped by broadcast_group_id of progress.
+        // Messages in msg_group will be pushed to message queue later.
         let core = &mut self.r;
         let msgs = &mut self.msgs;
-        self.prs
-            .iter_mut()
+        let prs = &mut self.prs.progress;
+        let conf = &self.prs.conf;
+        prs.iter_mut()
             .filter(|&(id, _)| *id != self_id)
-            .for_each(|(id, pr)| core.send_append(*id, pr, msgs));
+            .for_each(|(id, pr)| {
+                let mut tmp_msgs = Vec::default();
+                // Let messages be pushed into tmp_vec firstly.
+                core.send_append(*id, pr, &mut tmp_msgs);
+                for msg in tmp_msgs {
+                    // Filter out messages that need to be forwarded into msg_group.
+                    // Other messages are sent directly.
+                    if pr.broadcast_group_id == leader_group_id
+                        || msg.get_msg_type() != MessageType::MsgAppend
+                        || !pr.is_replicating()
+                    {
+                        msgs.push(msg);
+                    } else {
+                        let peer_id = msg.to;
+                        core.msg_group
+                            .entry(pr.broadcast_group_id)
+                            .or_default()
+                            // The agent must be a voter and active recently.
+                            .push((msg, pr.recent_active && conf.voters().contains(peer_id)));
+                    }
+                }
+            });
+
+        // Merge messages in the same broadcast group and send them.
+        for (_, mut group) in core.msg_group.drain() {
+            let mut need_merge = group.len() > 1;
+            let mut agent_msg_idx = None;
+            if need_merge {
+                // If found an appropriate agent, return the index of agent's message. Otherwise, return None.
+                agent_msg_idx = group
+                    .iter()
+                    .position(|(_, is_agent_candidate)| *is_agent_candidate);
+                need_merge = agent_msg_idx.is_some();
+            }
+            // Do not need to merge if group size is less than two. Or there is no appropriate agent.
+            if !need_merge {
+                msgs.append(&mut group.into_iter().map(|(msg, _)| msg).collect());
+                return;
```

Comment:
continue

Reply:
Thank very much. I think this cause the jitters of the number of sent messages.

Reply:
you should improve the test to cover this